Reduce use of CSSCalcTree.h and CalculationTree.h in header files

weinig opened the pull request:
#### 5c0ff0cb989fa04a1ee848c39073574024d91ae3
<pre>
Reduce use of CSSCalcTree.h and CalculationTree.h in header files
<a href="https://bugs.webkit.org/show_bug.cgi?id=286514">https://bugs.webkit.org/show_bug.cgi?id=286514</a>

Reviewed by Antti Koivisto.

Makes CSSCalc::Child, CSSCalc::ChildOrNone and CSSCalc::Children,
and Calculation counterparts, forward declarable by wrapping the
complex definitions in structs. This also reduces the number of
times some template types have to be instantiated.

Additionally moves mappings of CSSCalcTree to/from CalculationTree
out of CSSCalcTree.h and into its own header. That allows removing
inclusion of CalculationTree.h from CSSCalcTree.h, cutting down on
more include bloat.

With these types now forward declarable, replace unnecessary
inclusions of CSSCalcTree.h and CalculationTree.h with forward
declarations in headers.

* Source/WTF/wtf/StdLibExtras.h:
Replace strict const reference requirement with a universal reference
to allow switching on non-const variant-likes.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.h:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.h:
* Source/WebCore/css/calc/CSSCalcTree+Mappings.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.h:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.h:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/query/MediaQueryParser.h:
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
* Source/WebCore/css/typedom/CSSNumericValue.h:
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathInvert.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathMax.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathMin.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathNegate.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathProduct.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathSum.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathValue.cpp:
* Source/WebCore/platform/calc/CalculationOperator.h:
* Source/WebCore/platform/calc/CalculationRange.h: Copied from Source/WebCore/platform/calc/CalculationTree+Evaluation.h.
* Source/WebCore/platform/calc/CalculationTree+Evaluation.cpp:
* Source/WebCore/platform/calc/CalculationTree+Evaluation.h:
* Source/WebCore/platform/calc/CalculationTree.cpp:
* Source/WebCore/platform/calc/CalculationTree.h:
* Source/WebCore/platform/calc/CalculationValue.cpp:
* Source/WebCore/platform/calc/CalculationValue.h:
* Source/WebCore/style/values/primitives/StylePosition.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.h:

Canonical link: <a href="https://commits.webkit.org/289559@main">https://commits.webkit.org/289559@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/282c933008d2c4867e9bba017a4e14c94dcbd075

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87242 "79 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92103 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37982 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89292 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7028 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14824 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67424 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25162 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90244 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5391 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78959 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47746 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5164 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33338 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37098 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/80031 "Found 1651 new JSC stress test failures: basic-tests.yaml/stress-test.js.ftl-eager-no-cjit, cdjs-tests.yaml/main.js.lockdown, cdjs-tests.yaml/motion_test.js.ftl-eager-no-cjit, cdjs-tests.yaml/red_black_tree_test.js.ftl-eager-no-cjit, cdjs-tests.yaml/reduce_collision_set_test.js.ftl-eager-no-cjit, microbenchmarks/Float32Array-matrix-mult.js.ftl-eager-no-cjit, microbenchmarks/Int16Array-bubble-sort-with-byteLength.js.ftl-eager-no-cjit, microbenchmarks/JSONP-negative-0.js.ftl-eager-no-cjit, microbenchmarks/Number-isNaN.js.ftl-eager-no-cjit, microbenchmarks/abc-postfix-backward-loop.js.ftl-eager-no-cjit ... (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75648 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34214 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93988 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86016 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14404 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10489 "Found 1 new test failure: fast/css/view-transitions-zoom.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76232 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14608 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74815 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75437 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18572 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19779 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18216 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7333 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14423 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19716 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/108510 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14168 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26115 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17611 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15949 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/platform/calc/CalculationRange.h:
```diff
@@ -0,0 +1,43 @@
+namespace WebCore {
+namespace Calculation {
+
+struct Range {
+    double min;
+    double max;
+
+    constexpr bool operator==(const Range&) const = default;
+};
+
+inline constexpr auto All = Range { -std::numeric_limits<double>::infinity(), std::numeric_limits<double>::infinity() };
```
Maybe this could be `Range::All`?

weinig replied:
It can't for dumb c++ reasons. You can't declare a static variable of the type of the struct/class you are currently defining.

It could be a function that returns one, so `Range::All()`, but that's not great either.

weinig committed:
Reduce use of CSSCalcTree.h and CalculationTree.h in header files
https://bugs.webkit.org/show_bug.cgi?id=286514

Reviewed by Antti Koivisto.

Makes CSSCalc::Child, CSSCalc::ChildOrNone and CSSCalc::Children,
and Calculation counterparts, forward declarable by wrapping the
complex definitions in structs. This also reduces the number of
times some template types have to be instantiated.

Additionally moves mappings of CSSCalcTree to/from CalculationTree
out of CSSCalcTree.h and into its own header. That allows removing
inclusion of CalculationTree.h from CSSCalcTree.h, cutting down on
more include bloat.

With these types now forward declarable, replace unnecessary
inclusions of CSSCalcTree.h and CalculationTree.h with forward
declarations in headers.

* Source/WTF/wtf/StdLibExtras.h:
Replace strict const reference requirement with a universal reference
to allow switching on non-const variant-likes.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.h:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.h:
* Source/WebCore/css/calc/CSSCalcTree+Mappings.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.h:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.h:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/query/MediaQueryParser.h:
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
* Source/WebCore/css/typedom/CSSNumericValue.h:
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathInvert.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathMax.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathMin.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathNegate.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathProduct.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathSum.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathValue.cpp:
* Source/WebCore/platform/calc/CalculationOperator.h:
* Source/WebCore/platform/calc/CalculationRange.h: Copied from Source/WebCore/platform/calc/CalculationTree+Evaluation.h.
* Source/WebCore/platform/calc/CalculationTree+Evaluation.cpp:
* Source/WebCore/platform/calc/CalculationTree+Evaluation.h:
* Source/WebCore/platform/calc/CalculationTree.cpp:
* Source/WebCore/platform/calc/CalculationTree.h:
* Source/WebCore/platform/calc/CalculationValue.cpp:
* Source/WebCore/platform/calc/CalculationValue.h:
* Source/WebCore/style/values/primitives/StylePosition.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.h:

Canonical link: https://commits.webkit.org/289559@main
